diff --git a/app/controllers/omniauth_callbacks_controller.rb b/app/controllers/omniauth_callbacks_controller.rb
index 5e6676e..b884f0f 100644
--- a/app/controllers/omniauth_callbacks_controller.rb
+++ b/app/controllers/omniauth_callbacks_controller.rb
@@ -5,7 +5,14 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
   protect_from_forgery except: [:kerberos, :saml, :cas3]
 
   Gitlab.config.omniauth.providers.each do |provider|
-    define_method provider['name'] do
+    provider_name = provider['name']
+
+    # Allow for multiple OpenID Connect providers
+    if provider_name == 'openid_connect'
+        provider_name = provider['args']['name']
+    end
+
+    define_method provider_name do
       handle_omniauth
     end
   end
diff --git a/lib/gitlab/auth/ldap/person.rb b/lib/gitlab/auth/ldap/person.rb
index 8dfae3e..8ea1961 100644
--- a/lib/gitlab/auth/ldap/person.rb
+++ b/lib/gitlab/auth/ldap/person.rb
@@ -16,6 +16,11 @@ module Gitlab
           adapter.user(adapter.config.uid, uid)
         end
 
+        def self.find_by_uuid(uuid, adapter)
+          uuid = Net::LDAP::Filter.escape(uuid)
+          adapter.user('entryUUID', uuid)
+        end
+
         def self.find_by_dn(dn, adapter)
           adapter.user('dn', dn)
         end
diff --git a/lib/gitlab/auth/o_auth/user.rb b/lib/gitlab/auth/o_auth/user.rb
index b6a9608..699d07d 100644
--- a/lib/gitlab/auth/o_auth/user.rb
+++ b/lib/gitlab/auth/o_auth/user.rb
@@ -122,6 +122,7 @@ module Gitlab
 
         def find_ldap_person(auth_hash, adapter)
           Gitlab::Auth::LDAP::Person.find_by_uid(auth_hash.uid, adapter) ||
+            Gitlab::Auth::LDAP::Person.find_by_uuid(auth_hash.uid, adapter) ||
             Gitlab::Auth::LDAP::Person.find_by_email(auth_hash.uid, adapter) ||
             Gitlab::Auth::LDAP::Person.find_by_dn(auth_hash.uid, adapter)
         end
